Source/Core/Core/PowerPC/MMU.cpp

```diff
@@ -1181,19 +1105,136 @@ static __forceinline u32 TranslatePageAddress(const u32 address, const XCheckTLB
         if (res != TLB_UPDATE_C)
           UpdateTLBEntry(flag, PTE2, address);
 
-        return (PTE2.RPN << 12) | offset;
+        return TranslateAddressResult{TranslateAddressResult::PAGE_TABLE_TRANSLATED,
+                                      (PTE2.RPN << 12) | offset};
       }
     }
   }
-  return 0;
+  return TranslateAddressResult{TranslateAddressResult::PAGE_FAULT, 0};
+}
+
+static void UpdateBATs(u32* bat_table, u32 base_spr)
+{
+  // TODO: Separate BATs for MSR.PR==0 and MSR.PR==1
+  // TODO: Handle PP/WIMG settings.
+  // TODO: Check how hardware reacts to overlapping BATs (including
+  // BATs which should cause a DSI).
+  // TODO: Check how hardware reacts to invalid BATs (bad mask etc).
+  for (int i = 0; i < 4; ++i)
+  {
+    u32 spr = base_spr + i * 2;
+    UReg_BAT_Up batu = PowerPC::ppcState.spr[spr];
+    UReg_BAT_Lo batl = PowerPC::ppcState.spr[spr + 1];
+    if (batu.VS == 0 && batu.VP == 0)
+      continue;
+
+    if ((batu.BEPI & batu.BL) != 0)
+    {
+      // With a valid BAT, the simplest way to match is
+      // (input & ~BL_mask) == BEPI. For now, assume it's
+      // implemented this way for invalid BATs as well.
+      WARN_LOG(POWERPC, "Bad BAT setup: BEPI overlaps BL");
+      continue;
+    }
+    if ((batl.BRPN & batu.BL) != 0)
+    {
+      // With a valid BAT, the simplest way to translate is
+      // (input & BL_mask) | BRPN_address. For now, assume it's
+      // implemented this way for invalid BATs as well.
+      WARN_LOG(POWERPC, "Bad BAT setup: BPRN overlaps BL");
+    }
+    if (CountSetBits((u32)(batu.BL + 1)) != 1)
+    {
+      // With a valid BAT, the simplest way of masking is
+      // (input & ~BL_mask) for matching and (input & BL_mask) for
+      // translation. For now, assume it's implemented this way for
+      // invalid BATs as well.
+      WARN_LOG(POWERPC, "Bad BAT setup: invalid mask in BL");
+    }
+    for (u32 j = 0; j < (1 << 11); ++j)
```

Comment:
The optimized version is probably `for (u32 j = 0, mask = batu.BL; j <= mask; j = (j - mask) & mask)` a trick I learned from [ryg's Texture tiling and swizzling](https://fgiesen.wordpress.com/2011/01/17/texture-tiling-and-swizzling/) that's useful surprisingly often. In this case I'd go with your suggestion for clarity.